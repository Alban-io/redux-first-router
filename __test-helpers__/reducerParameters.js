import createSmartHistory from '../src/smart-history'

import { createInitialState } from '../src/createLocationReducer'

export default async (type, pathname) => {
  // eslint-disable-line import/prefer-default-export
  const history = createSmartHistory({ initialEntries: ['/first'] })
  history.listen(function() {})
  await history.push(pathname)

  const current = { pathname, type, payload: { param: 'bar' } }
  const prev = { pathname: '/first', type: 'FIRST', payload: {} }
  const routesMap = {
    FIRST: { path: '/first' },
    SECOND: { path: '/second/:param' }
  }

  return {
    type,
    pathname,
    current,
    prev,

    initialState: createInitialState(
      routesMap,
      history
    ),

    routesMap,
    history,

    action: {
      type,
      payload: { param: 'bar' },
      meta: {
        location: {
          current,
          prev,
          kind: 'load',
          history: {
            entries: history.entries.slice(0), // history.entries.map(entry => entry.pathname)
            index: history.index,
            length: history.length
          }
        }
      }
    },

    expectState(state) {
      expect(state.pathname).toEqual(pathname)
      expect(state.type).toEqual(type)
      expect(state.payload).toEqual({ param: 'bar' })
      expect(state.prev).toEqual(prev)
      expect(state.kind).toEqual('load')

      expect(state).toMatchSnapshot()
    }
  }
}
