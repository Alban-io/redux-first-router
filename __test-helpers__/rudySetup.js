import { applyMiddleware, createStore, compose, combineReducers } from 'redux'
import reduxThunk from 'redux-thunk'
import createRouter from '../src/createRouter'
import createScene from '../src/createScene'
import { NOT_FOUND } from '../src/index'

import fakeAsyncWork from '../__test-helpers__/fakeAsyncWork'

export default (path = '/first', options = {}, custom = {}) => {
  const routesMap = {
    ...defaultRoutes,
    ...custom.routesMap
  }

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { types, actions, routes, exportString } = createScene(routesMap, custom)
  const { middleware, reducer, firstRoute, history } = createRouter(
    routes,
    options
  )

  const title = (state = {}, action = {}) => action.type
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware, reduxThunk)
  const store = createStore(rootReducer, enhancer)

  return {
    store,
    firstRoute,
    history,
    types,
    actions,
    exportString,
    routes
  }
}

export const defaultRoutes = {
  FIRST: '/first',
  SECOND: '/second',
  THIRD: {
    path: '/third',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    }
  },
  FOURTH: {
    path: '/fourth',
    thunk: async () => {
      await fakeAsyncWork()
      return 'thunk'
    },
    onComplete: () => {
      return 'onComplete'
    },
    action: (arg) => (req, type) => {
      return { payload: { foo: arg }, type }
    }
  },
  PLAIN: {
    action: (arg) => {
      return { foo: arg }
    }
  },
  [NOT_FOUND]: '/not-found-foo'
}

export const log = store => {
  const state = store.getState().location
  delete state.routesMap
  delete state.hasSSR
  console.log(state)
}
