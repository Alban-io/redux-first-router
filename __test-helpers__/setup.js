import createHistory from 'history/createMemoryHistory'
import connectRoutes from '../src/connectRoutes'


export default (path = '/', options = { title: 'title', location: 'location' }) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param',
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  return connectRoutes(history, routesMap, options)
}
