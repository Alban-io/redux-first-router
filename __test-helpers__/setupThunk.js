import { createStore, applyMiddleware, compose, combineReducers } from 'redux'
import connectRoutes from '../src/connectRoutes'

export default (path = '/', thunkArg, opts, dispatchFirstRoute = true) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: { path: '/second/:param', thunk: thunkArg },
    THIRD: { path: '/third/:param' }
  }

  const options = { extra: { arg: 'extra-arg' }, initialEntries: path, ...opts }

  const { middleware, enhancer, reducer, history, firstRoute } = connectRoutes(
    routesMap,
    options
  )

  const rootReducer = combineReducers({
    location: reducer
  })

  const middlewares = applyMiddleware(middleware)
  const enhancers = compose(enhancer, middlewares)
  const store = createStore(rootReducer, enhancers)

  if (dispatchFirstRoute) store.dispatch(firstRoute())

  return { store, firstRoute, history }
}
