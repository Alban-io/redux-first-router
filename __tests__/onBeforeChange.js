import { setupAll } from '../__test-helpers__/setup'
import fakeAsyncWork from '../__test-helpers__/fakeAsyncWork'
import tempMock from '../__test-helpers__/tempMock'

it('calls onBeforeChange handler on route change', () => {
  const onBeforeChange = jest.fn()
  const { store } = setupAll('/first', { onBeforeChange })

  const action = { type: 'SECOND', payload: { param: 'bar' } }
  store.dispatch(action)

  expect(onBeforeChange).toHaveBeenCalled()
  expect(onBeforeChange.mock.calls[1][2].action).toMatchObject(action)
  expect(onBeforeChange.mock.calls[1][2].arg).toEqual('extra-arg')
})

it('if onBeforeChange dispatches redirect, route changes with kind === "redirect"', async () => {
  const onBeforeChange = jest.fn((dispatch, getState, { action }) => {
    if (action.type !== 'SECOND') return
    dispatch({ type: 'THIRD' })
  })

  const { store, history } = setupAll('/first', { onBeforeChange })
  await store.dispatch({ type: 'SECOND', payload: { param: 'bar' } })
  const { location } = store.getState()

  expect(location.kind).toEqual('redirect')
  expect(location.type).toEqual('THIRD')
  expect(history.entries.length).toEqual(2)
  expect(location).toMatchSnapshot()
  expect(onBeforeChange.mock.calls.length).toEqual(3)
})

it('onBeforeChange redirect on server results in 1 history entry', async () => {
  tempMock('../src/pure-utils/isServer', () => () => true)
  const { setupAll } = require('../__test-helpers__/setup')

  const onBeforeChange = jest.fn((dispatch, getState, { action }) => {
    if (action.type !== 'SECOND') return
    dispatch({ type: 'THIRD' })
  })

  const { store, history } = setupAll('/first', {
    onBeforeChange
  })

  await store.dispatch({ type: 'SECOND', payload: { param: 'bar' } })

  const { location } = store.getState()
  expect(history.entries.length).toEqual(1) // what we are testing for
  expect(location).toMatchSnapshot()
})

it('onBeforeChange delays route changes until onBeforeChange promise resolves', async () => {
  const onBeforeChange = jest.fn(async (dispatch, getState, { action }) => {
    if (action.type !== 'SECOND') return

    await fakeAsyncWork()
    dispatch({ type: 'NON_ROUTE_AUTH_ACTION' })
  })

  const routesMap = {
    FIRST: '/first',
    SECOND: {
      path: '/second',
      thunk: async (dispatch, getState) => {
        await fakeAsyncWork()
        return 'thunkReturn'
      }
    }
  }

  const { store } = setupAll('/first', { onBeforeChange }, { routesMap })

  const prom = store.dispatch({ type: 'SECOND' })
  expect(store.getState().location.type).toEqual('FIRST')

  const res = await prom
  expect(res).toEqual('thunkReturn')
  expect(store.getState().location.type).toEqual('SECOND')
  expect(store.getState().location.kind).toEqual('push')

  // VERY IMPORTANT:
  // there will be 4 dispatches, but only 2 calls to `onBeforeChange`:
  // - 1 initialDispatch in setupAll
  // - 2 externally by the user (SECOND, SOME_NON_ROUTE_TYPE_SUCH_AS_AUTH_STUFF)
  // - and 1 additional by the middleware to re-dispatch SECOND
  expect(onBeforeChange.mock.calls.length).toEqual(2)
})

it('onBeforeChange that returns a promise which redirects', async () => {
  const onBeforeChange = jest.fn(async (dispatch, getState, { action }) => {
    if (action.type !== 'SECOND') return

    await fakeAsyncWork()
    dispatch({ type: 'THIRD' })
  })

  const routesMap = {
    FIRST: '/first',
    SECOND: '/second',
    THIRD: {
      path: '/third',
      thunk: async (dispatch, getState) => {
        await fakeAsyncWork()
        return 'onBeforeChangePlusThunkReturn'
      }
    }
  }

  const { store } = setupAll('/first', { onBeforeChange }, { routesMap })

  expect(store.getState().location.type).toEqual('FIRST')

  const res = await store.dispatch({ type: 'SECOND' })
  expect(res).toEqual('onBeforeChangePlusThunkReturn')
  expect(store.getState().location.type).toEqual('THIRD')
  expect(store.getState().location.kind).toEqual('redirect')

  // VERY IMPORTANT:
  // we expect `onBeforeChange` to be called 3 times, because even the final/3rd
  // route redirected to must be filtered again. In the final route, there are no more
  // redirects so it passes the filtering.
  expect(onBeforeChange.mock.calls.length).toEqual(3)
})
