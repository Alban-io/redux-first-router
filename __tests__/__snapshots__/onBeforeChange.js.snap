// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if onBeforeChange dispatches redirect, route changes with kind === "redirect" 1`] = `
Object {
  "hasSSR": undefined,
  "history": undefined,
  "kind": "redirect",
  "pathname": "/third",
  "payload": Object {},
  "prev": Object {
    "kind": "push",
    "pathname": "/second/bar",
    "payload": Object {
      "param": "bar",
    },
    "type": "SECOND",
  },
  "routesMap": Object {
    "FIRST": "/first",
    "SECOND": "/second/:param",
    "THIRD": "/third",
  },
  "type": "THIRD",
}
`;

exports[`onBeforeChange redirect on server results in 1 history entry 1`] = `
Object {
  "hasSSR": true,
  "history": undefined,
  "kind": "redirect",
  "pathname": "/third",
  "payload": Object {},
  "prev": Object {
    "kind": "push",
    "pathname": "/second/bar",
    "payload": Object {
      "param": "bar",
    },
    "type": "SECOND",
  },
  "routesMap": Object {
    "FIRST": "/first",
    "SECOND": "/second/:param",
    "THIRD": "/third",
  },
  "type": "THIRD",
}
`;
